### PR TITLE
`drawable_layers`, `CircuitGraph` don't depend on `Operator` eq or hash

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
Operator equality and hash is being updated soon. This change is in preparation for that.

**Description of the Change:**
* Updated `drawable_layers` to create lists of operators for the layers instead of sets.
* Updated `CircuitGraph` to not rely on the Operator hash for accessing graph nodes/edges. Instead, a dictionary now stores the node indices keyed by the identity of the node operators, which can then be used to manipulate the graph.
* Updated `CircuitGraph` to stop putting copies of `StateMP` and `SampleMP` observables in the graph instead of the original measurement process. This makes `CircuitGraph.has_path` and other methods that rely on specific operator references in a tape to work as expected.

**Benefits:**
* No dependency on operator `__eq__` or `__hash__` from `drawer` module or `CircuitGraph`.
* The updates to `CircuitGraph` open up the possibility of `finite-diff` and `spsa` gradient methods to work with state/density matrix measurements.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
[sc-38493]

File: tests/docs/test_supported_confs.py [hunk 1]
```diff
@@ -445,6 +445,7 @@ def test_all_finitediff_nonstate(self, interface, return_type, shots, wire_specs
         else:
             grad = compute_gradient(x, interface, circuit, return_type)
 
+    @pytest.mark.xfail
```
Comment (timmysilv) [inline tests/docs/test_supported_confs.py:448]: can you add a reason to this?

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -126,6 +126,9 @@
 
 * The construction of the pauli representation for the `Sum` class is now faster.
   [(#4142)](https://github.com/PennyLaneAI/pennylane/pull/4142)
+* `qml.drawer.drawable_layers.drawable_layers` and `qml.CircuitGraph` have been updated to not rely on `Operator`
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:129]:
```suggestion

* `qml.drawer.drawable_layers.drawable_layers` and `qml.CircuitGraph` have been updated to not rely on `Operator`
```

**Review by timmysilv — APPROVED** [review]
awesome!

File: pennylane/circuit_graph.py [hunk 3]
```diff
@@ -110,26 +110,31 @@ def __init__(self, ops, obs, wires, par_info=None, trainable_params=None):
...
                 # Sampling without specifying wires is treated as sampling all wires
-                op = qml.sample(wires=meas_wires)
+                copy_op = qml.sample(wires=meas_wires)
+            else:
+                copy_op = op
 
             op.queue_idx = k  # store the queue index in the Operator
 
-            for w in op.wires:
+            for w in copy_op.wires:
```
Comment (albi3ro) [inline pennylane/circuit_graph.py:137]: This is fine as is, but one thing I was doing was:  
```
for w in self.wires if len(op.wires) == 0 else op.wires:
```

No need for `copy_op`. less code. Makes is so you can remove the whole `if-elif-else` block.

That particularly important though.

**Review by albi3ro — APPROVED** [review]
🚀 🎉 

Left a minor suggestion on circuit graph, but it's nothing to block approval.

**Pull request merged**